src/Chess/ChessBoard.hs

```diff
@@ -15,11 +16,114 @@ freshBoard = fenToChessBoard "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR"
+                                    ++ retrievePieceAt ('E',5) chessBoard 
+                                    ++ retrievePieceAt ('F',5) chessBoard 
+                                    ++ retrievePieceAt ('G',5) chessBoard 
+                                    ++ retrievePieceAt ('H',5) chessBoard 
+                        ++ "\n"
+                        ++ "(4)  "  ++ retrievePieceAt ('A',4) chessBoard
+                                    ++ retrievePieceAt ('B',4) chessBoard 
+                                    ++ retrievePieceAt ('C',4) chessBoard 
+                                    ++ retrievePieceAt ('D',4) chessBoard 
+                                    ++ retrievePieceAt ('E',4) chessBoard 
+                                    ++ retrievePieceAt ('F',4) chessBoard 
+                                    ++ retrievePieceAt ('G',4) chessBoard 
+                                    ++ retrievePieceAt ('H',4) chessBoard 
+                        ++ "\n"
+                        ++ "(3)  "   ++ retrievePieceAt ('A',3) chessBoard
+                                    ++ retrievePieceAt ('B',3) chessBoard 
+                                    ++ retrievePieceAt ('C',3) chessBoard 
+                                    ++ retrievePieceAt ('D',3) chessBoard 
+                                    ++ retrievePieceAt ('E',3) chessBoard 
+                                    ++ retrievePieceAt ('F',3) chessBoard 
+                                    ++ retrievePieceAt ('G',3) chessBoard 
+                                    ++ retrievePieceAt ('H',3) chessBoard 
+                        ++ "\n"
+                        ++ "(2)  "  ++ retrievePieceAt ('A',2) chessBoard
+                                    ++ retrievePieceAt ('B',2) chessBoard 
+                                    ++ retrievePieceAt ('C',2) chessBoard 
+                                    ++ retrievePieceAt ('D',2) chessBoard 
+                                    ++ retrievePieceAt ('E',2) chessBoard 
+                                    ++ retrievePieceAt ('F',2) chessBoard 
+                                    ++ retrievePieceAt ('G',2) chessBoard 
+                                    ++ retrievePieceAt ('H',2) chessBoard 
+                        
+                        ++ "\n"
+                        ++ "(1)  "  ++ retrievePieceAt ('A',1) chessBoard
+                                    ++ retrievePieceAt ('B',1) chessBoard 
+                                    ++ retrievePieceAt ('C',1) chessBoard 
+                                    ++ retrievePieceAt ('D',1) chessBoard 
+                                    ++ retrievePieceAt ('E',1) chessBoard 
+                                    ++ retrievePieceAt ('F',1) chessBoard 
+                                    ++ retrievePieceAt ('G',1) chessBoard 
+                                    ++ retrievePieceAt ('H',1) chessBoard 
+                        
+                        ++ "\n\n" 
```

Comment:
Note that if you're doing the same operation on all the positions on the chess board, you can use `foldl`.

You can create a list of all positions on the chessboard like this:
```haskell
allSquares = [(char, num) | char <- "ABCDEFGH", num <- [1, 2, 3, 4, 5, 6, 7, 8]]
```

Then you can do something like:
```haskell
stringChessBoard = foldl (\ position stringSoFar -> (retrievePieceAt position chessBoard)++stringSoFar) "" allSquares 
```
again, you'll have to test this, and also insert the newlines and numbers somehow. I still think it would be faster to replace substrings in a template. There is an implementation of replace [here](https://stackoverflow.com/questions/48492142/replace-a-substring-from-a-string-in-haskell)

If you had a template chessboard string with positions marked as `"A1"`, `"H5"`, etc, then you can do:
```haskell
-- assuming replace has signature: replace oldSubstring newSubstring wholeString
-- assuming (read position) returns something like (read ('A', 1)) = "A1"
stringChessBoard chessBoard = foldl (\ position templateString -> replace (read position) (retrievePieceAt position chessBoard) templateString) templateChessBoard allSquares
```

At the end of the day, as long as it works and the tests pass, implementation doesn't really matter that much